src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -7485,4 +7574,193 @@ bool StorageReplicatedMergeTree::createEmptyPartInsteadOfLost(zkutil::ZooKeeperP
+    }
+
+    void save(DiskPtr disk, const String & path) const
+    {
+        auto file_path = getFileName(path);
+        auto buffer = disk->writeMetaFile(file_path);
+        writeIntText(version, *buffer);
+        buffer->write("\n", 1);
+        writeBoolText(is_replicated, *buffer);
+        buffer->write("\n", 1);
+        writeBoolText(is_remote, *buffer);
+        buffer->write("\n", 1);
+        writeString(replica_name, *buffer);
+        buffer->write("\n", 1);
+        writeString(zookeeper_name, *buffer);
+        buffer->write("\n", 1);
+        writeString(table_uuid, *buffer);
+        buffer->write("\n", 1);
+    }
+
+    bool load(DiskPtr disk, const String & path)
+    {
+        auto file_path = getFileName(path);
+        if (!disk->exists(file_path))
+            return false;
+        auto buffer = disk->readMetaFile(file_path);
+        readIntText(version, *buffer);
+        if (version != 1)
+        {
+            LOG_ERROR(&Poco::Logger::get("FreezeMetaData"), "Unknown freezed metadata version: {}", version);
+            return false;
+        }
+        DB::assertChar('\n', *buffer);
+        readBoolText(is_replicated, *buffer);
+        DB::assertChar('\n', *buffer);
+        readBoolText(is_remote, *buffer);
+        DB::assertChar('\n', *buffer);
+        readString(replica_name, *buffer);
+        DB::assertChar('\n', *buffer);
+        readString(zookeeper_name, *buffer);
+        DB::assertChar('\n', *buffer);
+        readString(table_uuid, *buffer);
+        DB::assertChar('\n', *buffer);
+        return true;
+    }
+
+    static void clean(DiskPtr disk, const String & path)
+    {
+        disk->removeMetaFileIfExists(getFileName(path));
+    }
+
+private:
+    static String getFileName(const String & path)
+    {
+        return fs::path(path) / "frozen_metadata.txt";
+    }
+
+public:
+    int version = 1;
+    bool is_replicated;
+    bool is_remote;
+    String replica_name;
+    String zookeeper_name;
+    String table_uuid;
+};
+
+
+bool StorageReplicatedMergeTree::removeDetachedPart(DiskPtr disk, const String & path, const String & part_name, bool is_freezed)
+{
+    if (disk->supportZeroCopyReplication())
+    {
+        if (is_freezed)
+        {
+            FreezeMetaData meta;
+            if (meta.load(disk, path))
+            {
+                FreezeMetaData::clean(disk, path);
+                return removeSharedDetachedPart(disk, path, part_name, meta.table_uuid, meta.zookeeper_name, meta.replica_name, "");
+            }
+        }
+        else
+        {
+            String table_uuid = getTableUniqID();
+
+            return removeSharedDetachedPart(disk, path, part_name, table_uuid, zookeeper_name, replica_name, zookeeper_path);
+        }
+    }
+
+    disk->removeRecursive(path);
+
+    return false;
+}
+
+
+bool StorageReplicatedMergeTree::removeSharedDetachedPart(DiskPtr disk, const String & path, const String & part_name, const String & table_uuid,
```

Comment:
String & path? Path to what?

Reply:
Path to detached part on disk, like ".../table_name/detached/1_0_0_0/"